autrainer/models/abstract_model.py

```diff
@@ -14,8 +17,7 @@ def __init__(self, output_dim: int) -> None:
         super().__init__()
         self.output_dim = output_dim
 
-    @abstractmethod
-    def forward(self, x: torch.Tensor) -> torch.Tensor:
+    def forward(self, x: Union[Data, torch.Tensor]) -> torch.Tensor:
```

Comment:
Stupid me, I realize I don't need the `_forward` class since I can call `super().forward()`, I'll push an update shortly

Reply:
See 4e606be5a167f8f1f673ed9ee6e104ecca11c91f

Reply:
Okay i dont get how this will work can you explain please? The `AbstractModel` now has a super call?

Reply:
Exactly. It will call `super().forward(...)`, which begins to resolve right to left and stops at the call to `torch.nn.Module.forward(...)` since the other classes don't have a `forward(...)` method (if they had, that would be a bummer).

So then `AbstractModel` implements the parsing of the data (if it's `Data`, then pass `Data.features`, if it's `torch.Tensor`, then pass the raw input), and leaves the "actual" `forward(...)` implementation to the derived class.

---
P.S.

Now that I think about it with a bit more clear head, I need to double-check the inheritance resolution to make sure we don't end up in some sort of weird infinite recurrence loop and/or break backprop

Reply:
Forget what I said above, I am obviously wrong, I'll try another solution in a bit...

Reply:
Then, to contradict myself again, the whole idea behind this is to have an easy way to override the parsing for custom models. This would speak in favor of a `_parse_data` method that can be overriden by the user in their local code, and thus the current implementation. 

So to summarize the three options as I see them:

1. The user writes a `forward(self, torch.Tensor)` function as normal. `autrainer` wraps it in a `ModelWrapper` under-the-hood to take care of the unwrapping of data. The benefit is that the user doesn't have to care at all about this functionality. However, this makes the whole PR "too much ado about nothing" as the whole point from the start was to make this configurable and allow models to benefit from added entries to the `Data` struct. I would thus drop this option.

2. The user writes a `forward(self, Data)` function. They have to call `AbstractModel._parse_data(self, Data)` or else their model will crash (not sure how we can make this error message more informative). They also have the option to override `AbstractModel._parse_data(self, Data)` and integrate custom logic to utilize more fields of an extended `Data` struct.

3. Similarly, we can force the user to implement a private `_function(self, Tensor)` method and implement `AbstractModel.forward(self, Union[Data, Tensor])` to take care of the unwrapping. The downside of that, is that for custom logic that unpacks more stuff out of `Data`, the user would additionally need to reimplement the `forward(...)` function.

So, in summary, I think I lean towards 2. This is also the solution currently implemented in this branch. This solution sacrifice ease-of-use for extended functionality. I think it's worth it as I will "unlock" a lot more capabilities.

Reply:
Why do we not enforce a strict interface `forward(self, data: Data)` all the time? Since we get this from the Dataset, this feels like it makes the most sense to me if we want to pass full control to the user and let them do whatever they want with the inputs.
E.g. inherit from `Data` (provide lets say an abstract implementation of the bare minimum fields that we need internally), and then they could add their dataclasses on both the dataset and model side?

I feel like we loose the main benefit of moving to dataclasses when we still pass only the Tensor to the model.

Reply:
> Why do we not enforce a strict interface forward(self, data: Data) all the time?

Right now we do, actually, and provide the utility method `self._parse_data()` to handle the default "unpacking" of the struct. But the problem is that `python` does not "enforce" anything, so users coming from native `PyTorch` will have to adapt to the new way of doing things (which is fine, it's what `transformers` have done and it's been very successful there).

Reply:
While your IDE most likely will help you and generate the proper type annotations when you implement the abstract forward method, i definitely see your concerns and having `def forward(self, x: Union[Data, Tensor])->...` is non standard and inconvenient.

If i remember correctly then transformers also somewhat adhere to the PyTorch API and destructure their dataclasses on input?

This would also be a solution to the problem, such that we DO enforce some `standard` structure of the Data struct (+optionally also the forward function arguments).
We could check on the caller side which arguments of the current Data struct "fit in" the forward method and pass them directly as tensors, and enforce that the first one always has to the the features / inputs tensor (this would solve all our current issues). We even could go so far to not care if the first argument is named x, inputs, features, etc. and just always pass the features tensor this way. This is somewhat similar to the transformers implementation. Most forward methods have the other arguments as optional as well here (which we would need as well for inference at the moment).

Lets say a user defines a dataset that additionally returns a `segmentation_mask` attribute for a custom Data struct, we could similarly destructure it and pass it to the appropriate keyword argument.

Down the line, once we implement dataclasses support for transforms and augmentations, it could be the case that a transformer feature extractor transform returns an additional `attention_mask` we should pass to the model (e.g., creates a new Data struct with that field) and then we would handle it the same way we did above.

What do you think about this approach? Are there major drawbacks that i have missed?